src/lib/dma.c

```diff
@@ -47,6 +47,8 @@ struct dma *dma_get(uint32_t dir, uint32_t cap, uint32_t dev, uint32_t flags)
 	/* find DMAC with free channels that matches request */
 	for (d = lib_dma.dma_array; d < lib_dma.dma_array + lib_dma.num_dmas;
 	     d++) {
+		dcache_invalidate_region(d, sizeof(*d));
```

Comment:
Don't you have to write back cached then for this array elements? E.g. down in this function there's a
```
		dmin->sref++;
```
statement, which might be lost if you invalidate cache without writing it back to RAM first. Probably there are other occurrences too.
Wouldn't it be more efficient to invalidate the whole array at once? We do have one `break` statement in this loop, so if it's taken often and early, than invalidating individual elements might be faster.
Besides, are we sure this cannot corrupt adjacent data? The `struct dma` isn't cache-line aligned, so, say in `platform/intel/cavs/lib/dma.c` the array of DMA controllers can be followed by another data block in memory, not cache-line aligned. Then when you invalidate caches of the last dma array member, it will also invalidate a part of the following data block, potentially losing data.